go-controller/pkg/node/ovspinning/ovspinning_linux_test.go

```diff
@@ -0,0 +1,187 @@
+//go:build linux
+// +build linux
+
+package ovspinning
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"runtime"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/sys/unix"
+	"k8s.io/klog/v2"
+)
+
+func TestAlignCPUAffinity(t *testing.T) {
+
+	ovsDBPid, ovsDBStop := mockOvsdbProcess(t)
+	defer ovsDBStop()
+
+	ovsVSwitchdPid, ovsVSwitchdStop := mockOvsVSwitchdProcess(t)
+	defer ovsVSwitchdStop()
+
+	defer setTickDuration(20 * time.Millisecond)()
+	defer mockFeatureEnableFile(t, "1")()
+
+	var wg sync.WaitGroup
+	stopCh := make(chan struct{})
+	defer func() {
+		close(stopCh)
+		wg.Wait()
+	}()
+
+	wg.Add(1)
+	go func() {
+		// Be sure the system under test goroutine is finished before cleaning
+		defer wg.Done()
+		Run(stopCh)
+	}()
+
+	var initialCPUset unix.CPUSet
+	err := unix.SchedGetaffinity(os.Getpid(), &initialCPUset)
+	assert.NoError(t, err)
+
+	defer func() {
+		// Restore any previous CPU affinity value it was in place before the test
+		err = unix.SchedSetaffinity(os.Getpid(), &initialCPUset)
+		assert.NoError(t, err)
+	}()
+
+	assert.Greater(t, runtime.NumCPU(), 1)
+
+	for i := 0; i < runtime.NumCPU(); i++ {
+		var tmpCPUset unix.CPUSet
+		tmpCPUset.Set(i)
+		err = unix.SchedSetaffinity(os.Getpid(), &tmpCPUset)
+		assert.NoError(t, err)
+
+		klog.Infof("Test CPU Affinity %x", tmpCPUset)
+
+		assertPIDHasSchedAffinity(t, ovsVSwitchdPid, tmpCPUset)
+		assertPIDHasSchedAffinity(t, ovsDBPid, tmpCPUset)
+	}
+
+	// Disable the feature by making the enabler file empty
+	ioutil.WriteFile(featureEnablerFile, []byte(""), 0)
+	assert.NoError(t, err)
+
+	var tmpCPUset unix.CPUSet
```

Comment:
I would add a comment stating what you're after: change ovnk node cpuset and verify that, without the enabler file, ovs vswitchd and db server are not moved to the new cpuset for ovnk node.

I have started to wonder about this whole "disabling" thing... in these code changes, once the feature is enabled, we can disable it only once, otherwise we have to restart ovnk. All good. But disabling the feature means that we simply stop aligning the cpuset of OVS with that of ovnk node. As a consequence, OVS will keep running on the cpuset on which the routine last moved it before exiting.  @trozet do we care about moving OVS processes back to where they were at startup? This would add more complexity and I see how we might want to skip that, but I thought it might be worth asking :)

Reply:
Good point. Stopping to align OVS affinity after the daemon has spread out of system-reserved CPUs can break guaranteed workloads.

Reply:
Nice catch Ricky. So do we need to taskset it back to the reserved cpus? How would we know what those are? Look at what cpus kubelet has?

Reply:
The bad scenario is something like the following (correct me if I'm wrong):
- Suppose the node has 10 CPU and `systemReservedCPUs = 0,1`
- Starting with no guaranteed pods, onkube-node CPU affinity is `0-9` (it also has 0-1 for [kubernetes#115994](https://github.com/kubernetes/kubernetes/issues/115994))
- Affinity pinning routine set ovs-vwitchd affinity to `0-9`
- The user removes (or makes empty) `/etc/opensvwitchd/enable_cpu_affinity_pinning` file
- A new guaranteed Pod spawns and get assigned CPU 7
- k8sCPUManager changes ovnkube-node affinity to `0-6,8-9`
- ovs-vwitchd can still use CPU 7.

The scenario does not involve any ovnkube-node crashes, and if I correctly understood Martin's point, we can accept this behavior and document it.

*Bonus point*: if most of the changes to the `enable_cpu_affinity_pinning` file come with a node reboot, we can avoid watching the filesystem for changes and simplify the routine.

Reply:
So the proper changes to /etc/opensvwitchd/enable_cpu_affinity_pinning file come with a reboot. And there is no issue there. The feature will be enabled or disabled before any pods are started.

The watch is useful for manual intervention and I think we can document that part of this emergency procedure is also the manual affinity update. We can be nicer of course and set a base affinity in case the user forgets (reserved cpus would mimic what we have now).

Reply:
Btw, is OVN upgrade linked to OCP? Because I assume it might cause a short (hopefully) outage of ovnkube-node too, right?

Reply:
Yeah ovnkube-node rolls out with upgrade to OCP, but the data plane continues to function so there should be no existing outage there. The only outage is to the control plane temporarily.

I agree with the user should just have to fix the taskset. If the user is intelligent enough to know they need to disable the pinning, then they probably also know what their reserved cpus are, and are capable of fixing the taskset after manual intervention to disable the pinning.

Reply:
@MarSik out of curiosity, are the system-reserved CPUs (the ones in use by OVS at startup) subject to change before a reboot?

Anyway, as long as we document this manual procedure, we should be good, as @trozet said. We can also implement a "restore" step later on if requested by customers. 
Now... where should this be documented?

Reply:
OVS is not going to be using just the system reserved cpus at startup. It will have access to all of them.